lib/Parse/ParseDecl.cpp

```diff
@@ -4034,13 +4034,20 @@ ParserStatus Parser::parseTypeAttribute(TypeAttributes &Attributes,
       if (justChecking) return makeParserError();
 
       // If this is the first attribute, and if we are on a simple decl, emit a
-      // fixit to move the attribute.  Otherwise, we don't have the location of
-      // the @ sign, or we don't have confidence that the fixit will be right.
+      // fixit to move or just remove the attribute. Otherwise, we don't have
+      // the location of the @ sign, or we don't have confidence that the fixit
+      // will be right.
       if (!Attributes.empty() || StructureMarkers.empty() ||
-          StructureMarkers.back().Kind != StructureMarkerKind::Declaration ||
           StructureMarkers.back().Loc.isInvalid() ||
           peekToken().is(tok::equal)) {
         diagnose(Tok, diag::decl_attribute_applied_to_type);
+      } else if (InBindingPattern != PatternBindingState::NotInBinding ||
+                 StructureMarkers.back().Kind !=
+                     StructureMarkerKind::Declaration) {
```

Comment:
What are the cases where the last structure marker is not a declaration? AFAICT in all of your test cases I think the last structure marker should be a declaration but I might be wrong.

---

Also: What diagnostic do we produce now if the attribute is used in a generic position, e.g. 

```swift
struct Foo<A> {}
let a: Foo<@discardableResult () -> Int>
```

Reply:
> AFAICT in all of your test cases I think the last structure marker should be a declaration but I might be wrong.

The last two `if let _: @discardableResult Int = a { }` give me `(swift::Parser::StructureMarkerKind) $3 = Statement`

> Also: What diagnostic do we produce now if the attribute is used in a generic position, e.g.

Diagnostic keep the same, it is just be more conservative when offering the move it fix-it in such cases offer only to remove it makes more sense.